### PR TITLE
Optimize PartitionedOutputOperator - part 2

### PR DESCRIPTION
**Pull request by yingsu00**
The commits in this PR were cherry-picked from https://github.com/prestodb/presto/pull/13183 and were reviewed there.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by yingsu00** [comment]
@wenleix @arhimondr Wenlei and Andrii, thank you very much for reviewing PR https://github.com/prestodb/presto/pull/13183! I cherry-picked several standalone commits from that PR to this one. Your comments in #13183 have been addressed. Do you still have comments? If not, can we merge them first? Thank you!

File: presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java [hunk 1]
```diff
@@ -542,4 +624,153 @@ public static RunLengthEncodedBlock createRLEBlock(String value, int positionCou
...
         return new RunLengthEncodedBlock(blockBuilder.build(), positionCount);
     }
+
+    public static RunLengthEncodedBlock createRleBlockWithRandomValue(Block block, int positionCount)
+    {
+        checkArgument(block.getPositionCount() > 0, format("block positions %d is less than or equal to 0", block.getPositionCount()));
+        return new RunLengthEncodedBlock(block.getRegion(block.getPositionCount() / 2, 1), positionCount);
+    }
+
+    public static DictionaryBlock createDictionaryBlock(Block dictionary, int positionCount)
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java]: nit: `createRandomDictionaryBlock`?

File: presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java [hunk 2]
```diff
@@ -542,4 +624,153 @@ public static RunLengthEncodedBlock createRLEBlock(String value, int positionCou
...
+    {
+        int[] ids = IntStream.range(0, positionCount).map(i -> ThreadLocalRandom.current().nextInt(dictionary.getPositionCount() / 10)).toArray();
+        return new DictionaryBlock(dictionary, ids);
+    }
+
+    public static Block createBlockForType(Type type, int positionCount, boolean allowNulls, boolean createView, List<Encoding> wrappings)
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java]: nit: `createRandomBlockForType`?

File: presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java [hunk 3]
```diff
@@ -542,4 +624,153 @@ public static RunLengthEncodedBlock createRLEBlock(String value, int positionCou
...
+    {
+        DICTIONARY,
+        RUN_LENGTH
+    }
+
+    private static String generateRandomStringWithLength(int length)
+    {
+        byte[] array = new byte[length];
+        ThreadLocalRandom.current().nextBytes(array);
+        return new String(array, Charset.forName("UTF-8"));
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java]: `StandardCharsets.UTF_8` (static import `UTF_8`)

File: presto-main/src/test/java/com/facebook/presto/operator/PageAssertions.java [hunk 4]
```diff
@@ -36,4 +49,94 @@ public static void assertPageEquals(List<? extends Type> types, Page actualPage,
             assertBlockEquals(types.get(i), actualPage.getBlock(i), expectedPage.getBlock(i));
         }
     }
+
+    public static Page createPage(List<Type> types, int positionCount, boolean allowNulls)
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/operator/PageAssertions.java]: nit: `createPageWithRandomData`?

File: presto-main/src/test/java/com/facebook/presto/operator/PageAssertions.java [hunk 5]
```diff
@@ -36,4 +49,94 @@ public static void assertPageEquals(List<? extends Type> types, Page actualPage,
...
+    {
+        return createPage(types, positionCount, true, false, allowNulls, false, ImmutableList.of());
+    }
+
+    public static Page createDictionaryPage(List<Type> types, int positionCount, boolean allowNulls)
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/operator/PageAssertions.java]: nit: `createDictionaryPageWithRandomData`?

File: presto-main/src/test/java/com/facebook/presto/operator/PageAssertions.java [hunk 6]
```diff
@@ -36,4 +49,94 @@ public static void assertPageEquals(List<? extends Type> types, Page actualPage,
...
+    {
+        return createPage(types, positionCount, true, false, allowNulls, false, ImmutableList.of(DICTIONARY));
+    }
+
+    public static Page createRLEPage(List<Type> types, int positionCount, boolean allowNulls)
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/operator/PageAssertions.java]: nit: `createRlePageWithRandomData`?

File: presto-main/src/test/java/com/facebook/presto/operator/PageAssertions.java [hunk 7]
```diff
@@ -36,4 +49,94 @@ public static void assertPageEquals(List<? extends Type> types, Page actualPage,
...
+    {
+        return createPage(types, positionCount, true, false, allowNulls, false, ImmutableList.of(RUN_LENGTH));
+    }
+
+    public static Page createPage(
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/operator/PageAssertions.java]: nit: `createPageWithRandomData`?

**Comment by yingsu00** [comment]
@arhimondr Hi Andrii, the comments have all been addressed. Thanks very much for taking such careful look! @wenleix Do you have any more comments?
If it looks alright to you guys, could any of you help merge it?

**Comment by wenleix** [comment]
This is also previously approved by @wenleix  in #13183

**Pull request merged**